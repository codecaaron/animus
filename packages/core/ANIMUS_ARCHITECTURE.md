# Animus.ts Architecture Explanation

## Core Concept

Animus is a **type-state machine** that guides you through building styled components step-by-step. It uses TypeScript's type system to enforce that you configure your components in a specific order, making it impossible to make mistakes.

## The Unique Architecture

The most interesting aspect is the **backwards inheritance chain**:

```
Animus → AnimusWithBase → AnimusWithVariants → AnimusWithStates → AnimusWithSystem → AnimusWithAll
```

Each class extends the one *after* it in the chain, which seems backwards but is actually brilliant. Here's why:

- `Animus` (the starting point) inherits from `AnimusWithBase`
- `AnimusWithBase` inherits from `AnimusWithVariants`
- And so on...

This means the first class has access to all the methods defined in later classes, but those methods only become *available* to use when you reach the right stage.

## How It Works - Step by Step

### 1. Starting Point: `Animus` (lines 331-349)
```typescript
const builder = new Animus(propRegistry, groupRegistry);
```
- Takes your prop definitions and group configurations
- Only exposes the `.styles()` method
- Creates a parser for your props automatically

### 2. After calling `.styles()`: `AnimusWithBase` (lines 287-329)
```typescript
builder.styles({ padding: '20px', color: 'blue' })
```
- Now you can call `.variant()` to add design variations
- The base styles are locked in

### 3. After calling `.variant()`: `AnimusWithVariants` (lines 226-285)
```typescript
.variant({
  prop: 'size',
  variants: {
    small: { fontSize: '14px' },
    large: { fontSize: '20px' }
  }
})
```
- Can add multiple variants
- Now `.states()` method becomes available

### 4. After calling `.states()`: `AnimusWithStates` (lines 184-224)
```typescript
.states({
  disabled: { opacity: 0.5, cursor: 'not-allowed' },
  loading: { cursor: 'wait' }
})
```
- Defines boolean states (props that are true/false)
- Unlocks `.groups()` method

### 5. After calling `.groups()`: `AnimusWithSystem` (lines 140-182)
```typescript
.groups({ spacing: true, colors: true })
```
- Enables predefined groups of props
- Now you can add custom props with `.props()`

### 6. Final Stage: `AnimusWithAll` (lines 24-138)
This is where the magic happens. You can now:

- `.asElement('button')` - Create a styled HTML element
- `.asComponent(MyComponent)` - Style an existing React component  
- `.build()` - Get the raw styling function
- `.extend()` - Create variations of your styled component

## Key Features

### Type Safety Throughout
The 8 generic parameters track everything:
- `PropRegistry` - All your prop definitions
- `GroupRegistry` - Prop groupings
- `BaseParser` - How props convert to CSS
- `BaseStyles` - Your base CSS
- `Variants` - Design variations
- `States` - Boolean state styles
- `ActiveGroups` - Which groups are enabled
- `CustomProps` - Additional custom props

### Smart Prop Forwarding (lines 76-87)
```typescript
shouldForwardProp: (prop) => 
  isPropValid(prop) && !propNames.includes(prop)
```
- System props (like `spacing`, `colors`) are consumed and turned into styles
- Valid HTML attributes are passed through
- Invalid props are filtered out

### The Build Process (lines 99-137)
When you call `.build()`, it:
1. Merges your parser config with custom props
2. Creates a stylist that combines all your configurations
3. Returns a function that generates CSS objects
4. Handles variants, states, and system props automatically

### Extension System
The `.extend()` method creates an `AnimusExtended` instance that can break the rules - you can modify any part of the configuration in any order. This provides an escape hatch when needed.

## Why This Design?

1. **Impossible to Misuse** - You can't call methods in the wrong order
2. **Perfect IntelliSense** - Your IDE knows exactly what methods are available at each step
3. **Type Safe** - All configurations are fully typed
4. **Flexible When Needed** - The extend system provides flexibility
5. **Clean API** - The forced order mirrors how we think about styling (base → variants → states → props)

## Example Usage

```typescript
const Button = animus
  .styles({ 
    padding: '10px 20px',
    borderRadius: '4px' 
  })
  .variant({
    prop: 'variant',
    variants: {
      primary: { background: 'blue', color: 'white' },
      secondary: { background: 'gray', color: 'black' }
    }
  })
  .states({
    disabled: { opacity: 0.6 }
  })
  .asElement('button');

// Usage:
<Button variant="primary" disabled>Click me</Button>
```

## The Backwards Inheritance Explained

The backwards inheritance pattern is the key innovation. Instead of:
- Base class with few methods → Derived class with more methods (traditional OOP)

Animus does:
- Starting class that inherits everything → But methods are selectively exposed

This creates a "progressive revelation" pattern where:
1. You start with minimal options (just `.styles()`)
2. Each method call returns a new class instance
3. That new instance exposes the next logical methods
4. Invalid sequences are impossible at the type level

## Integration with Emotion

Animus builds on top of Emotion (a CSS-in-JS library):
- Uses `@emotion/styled` to create the actual styled components
- Uses `@emotion/is-prop-valid` to filter props intelligently
- The final output is a standard Emotion styled component
- Full compatibility with themes, SSR, and other Emotion features

## Performance Considerations

- All type checking happens at compile time - zero runtime overhead
- The inheritance chain is resolved during build
- Final components are standard Emotion components
- Parser and stylist are created once and reused

## Summary

This architecture ensures that every styled component is configured correctly, with all the power of CSS-in-JS but none of the footguns! The type-state machine pattern combined with backwards inheritance creates a unique developer experience where the API guides you to success.