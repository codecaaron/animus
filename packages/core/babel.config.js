module.exports = {
  extends: "../../babel.config.js",
  presets: ["@babel/preset-typescript"],
  include: ["./src/**/*.ts", "./src/**/*.tsx"],
  ignore: ["./**/*.d.ts", "__tests__"],
};
