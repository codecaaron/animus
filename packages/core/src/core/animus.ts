import { merge } from 'lodash';

import { Parser, Prop, SystemProps, TransformerMap } from '../types/config';
import {
  AbstractProps,
  CSSPropMap,
  CSSProps,
  CSSObject,
  ThemeProps,
} from '../types/props';

import { Arg, PropConfig } from '../types/config';
import { createStylist } from '../styles/createStylist';
import { create } from './create';
import styled from '@emotion/styled';

export class AnimusWithAll<
  C extends PropConfig,
  P extends Parser<TransformerMap<C['props']>>,
  Base extends CSSProps<AbstractProps, SystemProps<P>>,
  Variants extends Record<
    string,
    {
      prop?: any;
      defaultVariant?: any;
      base?: CSSProps<AbstractProps, SystemProps<P>>;
      variants: CSSPropMap<AbstractProps, SystemProps<P>>;
    }
  >,
  States extends CSSPropMap<AbstractProps, SystemProps<P>>,
  G extends (keyof C['groups'])[] | never[]
> {
  props = {} as C;
  parser = {} as P;
  groups = [] as G;
  base = {} as Base;
  statesConfig = {} as States;
  variants = {} as Variants;

  constructor(
    props: C,
    parser: P,
    base: Base,
    variants: Variants,
    states: States,
    groups: G
  ) {
    this.props = props;
    this.parser = parser;
    this.groups = groups;
    this.base = base;
    this.statesConfig = states;
    this.variants = variants;
  }

  asComponent<T extends keyof JSX.IntrinsicElements>(component: T) {
    type Props = ThemeProps<
      {
        [K in keyof Arg<P>]?: Arg<P>[K];
      } & { [K in keyof Variants]?: keyof Variants[K]['variants'] } & {
        [K in keyof States]?: boolean;
      }
    >;

    const handler = createStylist(
      this.parser,
      this.base,
      this.variants,
      this.statesConfig,
      {}
    ) as (props: { [K in keyof Props]: Props[K] }) => CSSObject;

    return styled(component)(handler);
  }

  build() {
    type Props = ThemeProps<
      {
        [K in keyof Arg<P>]?: Arg<P>[K];
      } & { [K in keyof Variants]?: keyof Variants[K]['variants'] } & {
        [K in keyof States]?: boolean;
      }
    >;

    const handler = createStylist(
      this.parser,
      this.base,
      this.variants,
      this.statesConfig,
      {}
    ) as (props: Props) => CSSObject;

    return handler;
  }
}

class AnimusWithSystem<
  C extends PropConfig,
  P extends Parser<TransformerMap<C['props']>>,
  Base extends CSSProps<AbstractProps, SystemProps<P>>,
  Variants extends Record<
    string,
    {
      prop?: any;
      defaultVariant?: any;
      base?: CSSProps<AbstractProps, SystemProps<P>>;
      variants: CSSPropMap<AbstractProps, SystemProps<P>>;
    }
  >,
  States extends CSSPropMap<AbstractProps, SystemProps<P>>,
  G extends (keyof C['groups'])[] | never[]
> extends AnimusWithAll<C, P, Base, Variants, States, G> {
  constructor(
    props: C,
    parser: P,
    base: Base,
    variants: Variants,
    states: States,
    groups: G
  ) {
    super(props, parser, base, variants, states, groups);
  }

  customProps<ExtraProps extends Record<string, Prop>>(config: ExtraProps) {
    const newProps = { ...this.props.props, ...config } as C['props'] &
      ExtraProps;
    const parser = create(newProps);
    return new AnimusWithAll(
      { ...this.props, props: newProps },
      parser,
      this.base,
      this.variants,
      this.statesConfig,
      this.groups
    );
  }
}

class AnimusWithStates<
  C extends PropConfig,
  P extends Parser<TransformerMap<C['props']>>,
  Base extends CSSProps<AbstractProps, SystemProps<P>>,
  Variants extends Record<
    string,
    {
      prop?: any;
      defaultVariant?: any;
      base?: CSSProps<AbstractProps, SystemProps<P>>;
      variants: CSSPropMap<AbstractProps, SystemProps<P>>;
    }
  >,
  States extends CSSPropMap<AbstractProps, SystemProps<P>>
> extends AnimusWithSystem<C, P, Base, Variants, States, []> {
  constructor(
    props: C,
    parser: P,
    base: Base,
    variants: Variants,
    states: States
  ) {
    super(props, parser, base, variants, states, []);
  }

  systemProps<PickedGroups extends keyof C['groups']>(
    config: Record<PickedGroups, true>
  ) {
    return new AnimusWithSystem(
      this.props,
      this.parser,
      this.base,
      this.variants,
      this.statesConfig,
      Object.keys(config) as PickedGroups[]
    );
  }
}

class AnimusWithVariants<
  C extends PropConfig,
  P extends Parser<TransformerMap<C['props']>>,
  Base extends CSSProps<AbstractProps, SystemProps<P>>,
  Variants extends Record<
    string,
    {
      prop?: any;
      defaultVariant?: any;
      base?: CSSProps<AbstractProps, SystemProps<P>>;
      variants: CSSPropMap<AbstractProps, SystemProps<P>>;
    }
  >
> extends AnimusWithStates<C, P, Base, Variants, {}> {
  constructor(props: C, parser: P, base: Base, variants: Variants) {
    super(props, parser, base, variants, {});
  }

  states<Props extends AbstractProps>(
    config: CSSPropMap<Props, SystemProps<P>>
  ) {
    return new AnimusWithStates(
      this.props,
      this.parser,
      this.base,
      this.variants,
      config
    );
  }

  variant<
    Keys extends keyof Props,
    Base extends AbstractProps,
    Props extends Record<Keys, AbstractProps>,
    PropKey extends Readonly<string> = 'variant'
  >(options: {
    prop?: PropKey;
    defaultVariant?: keyof Props;
    base?: CSSProps<Base, SystemProps<P>>;
    variants: CSSPropMap<Props, SystemProps<P>>;
  }) {
    type NextVariants = Variants & Record<PropKey, typeof options>;
    const prop = options.prop || 'variant';

    return new AnimusWithVariants(
      this.props,
      this.parser,
      this.base,
      merge(this.variants, { [prop]: options }) as NextVariants
    );
  }
}

class AnimusWithBase<
  C extends PropConfig,
  P extends Parser<TransformerMap<C['props']>>,
  Base extends CSSProps<AbstractProps, SystemProps<P>>
> extends AnimusWithVariants<C, P, Base, {}> {
  constructor(props: C, parser: P, base: Base) {
    super(props, parser, base, {});
  }
  variant<
    Keys extends keyof Props,
    Base extends AbstractProps,
    Props extends Record<Keys, AbstractProps>,
    PropKey extends Readonly<string> = 'variant'
  >(options: {
    prop?: PropKey;
    defaultVariant?: keyof Props;
    base?: CSSProps<Base, SystemProps<P>>;
    variants: CSSPropMap<Props, SystemProps<P>>;
  }) {
    type NextVariants = Record<PropKey, typeof options>;
    const prop = options.prop || 'variant';

    return new AnimusWithVariants(
      this.props,
      this.parser,
      this.base,
      merge(this.variants, { [prop]: options }) as NextVariants
    );
  }
}

export class Animus<
  C extends PropConfig,
  P extends Parser<TransformerMap<C['props']>>
> extends AnimusWithBase<C, P, {}> {
  constructor(config: C) {
    super(config, create(config.props) as P, {});
  }
  styles<Props extends AbstractProps>(config: CSSProps<Props, SystemProps<P>>) {
    return new AnimusWithBase(this.props, this.parser, config);
  }
}
