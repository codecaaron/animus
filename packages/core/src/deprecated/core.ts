import { createCss } from '../internal/createCss';
import { createVariant } from '../internal/createVariant';
import { create } from '../internal/create';
import { compose } from '../internal/compose';

export const animusProps = {
  compose,
  create,
  createCss,
  createVariant,
};
