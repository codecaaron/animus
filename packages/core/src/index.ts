export * from './types/props';
export * from './types/theme';
export * from './types/utils';
export * from './types/shared';

export * from './scales/createScale';
export * from './AnimusConfig';
export * from './createAnimus';
export * from './transforms';
export * from './legacy/core';

export { config } from './config';
import { config } from './config';

export const animus = config.build();
animus
  .styles({
    size: '2px',
  })
  .variant({
    prop: 'cool',
    variants: {
      many: {
        fontSize: 14,
      },
    },
  });
