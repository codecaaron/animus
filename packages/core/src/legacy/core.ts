import { createCss } from './createCss';
import { createVariant } from './createVariant';
import { create } from './create';
import { compose } from './compose';
import { createStates } from './createStates';

export const animusProps = {
  compose,
  create,
  createCss,
  createVariant,
  createStates,
};
