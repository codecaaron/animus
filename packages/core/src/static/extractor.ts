import * as parser from '@babel/parser';
// Handle the babel/traverse CommonJS export issue
// @ts-ignore - babel/traverse has complex module exports
import traverseDefault from '@babel/traverse';
import * as t from '@babel/types';

const traverse = (traverseDefault as any).default || traverseDefault;

import type { NodePath } from '@babel/traverse';

/**
 * Extracted style data from an Animus component
 */
export interface ExtractedStyles {
  componentName?: string;
  baseStyles?: Record<string, any>;
  variants?: Record<string, any> | Record<string, any>[];
  states?: Record<string, any>;
  groups?: string[];
  props?: Record<string, any>;
}

/**
 * Extract styles from Animus method chains in code
 */
export function extractStylesFromCode(code: string): ExtractedStyles[] {
  const ast = parser.parse(code, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  });

  const extractedComponents: ExtractedStyles[] = [];

  traverse(ast as any, {
    CallExpression(path: NodePath<t.CallExpression>) {
      // Look for .styles() calls
      if (
        t.isMemberExpression(path.node.callee) &&
        t.isIdentifier(path.node.callee.property, { name: 'styles' })
      ) {
        const extracted: ExtractedStyles = {};

        // Try to get component name from variable declaration
        const varDeclarator = path.findParent((p: NodePath) =>
          p.isVariableDeclarator()
        );
        if (varDeclarator && t.isVariableDeclarator(varDeclarator.node)) {
          if (t.isIdentifier(varDeclarator.node.id)) {
            extracted.componentName = varDeclarator.node.id.name;
          }
        }

        // Extract styles object from first argument
        const stylesArg = path.node.arguments[0];
        if (t.isObjectExpression(stylesArg)) {
          extracted.baseStyles = extractObjectLiteral(stylesArg);
        }

        // Look for chained method calls by traversing up the AST
        // We need to find the full chain starting from .styles()
        const chainedCalls: any[] = [];
        let currentPath: any = path;

        // Traverse up to find all chained calls
        while (currentPath) {
          if (
            t.isCallExpression(currentPath.node) &&
            t.isMemberExpression(currentPath.node.callee) &&
            t.isIdentifier(currentPath.node.callee.property)
          ) {
            chainedCalls.push({
              method: currentPath.node.callee.property.name,
              args: currentPath.node.arguments,
            });
          }

          // Check if parent is also a call expression (chained)
          if (
            currentPath.parentPath &&
            t.isMemberExpression(currentPath.parentPath.node) &&
            currentPath.parentPath.parentPath &&
            t.isCallExpression(currentPath.parentPath.parentPath.node)
          ) {
            currentPath = currentPath.parentPath.parentPath;
          } else {
            break;
          }
        }

        // Process chained calls in order
        const variantCalls: any[] = [];

        for (const call of chainedCalls) {
          if (call.method === 'variant' && call.args[0]) {
            const variantArg = call.args[0];
            if (t.isObjectExpression(variantArg)) {
              variantCalls.push(extractObjectLiteral(variantArg));
            }
          } else if (call.method === 'states' && call.args[0]) {
            const statesArg = call.args[0];
            if (t.isObjectExpression(statesArg)) {
              extracted.states = extractObjectLiteral(statesArg);
            }
          } else if (call.method === 'groups' && call.args[0]) {
            const groupsArg = call.args[0];
            if (t.isObjectExpression(groupsArg)) {
              const groups = extractObjectLiteral(groupsArg);
              extracted.groups = Object.keys(groups).filter(
                (key) => groups[key]
              );
            }
          } else if (call.method === 'props' && call.args[0]) {
            const propsArg = call.args[0];
            if (t.isObjectExpression(propsArg)) {
              extracted.props = extractObjectLiteral(propsArg);
            }
          }
        }

        // Store variants - if multiple, keep as array; if single, flatten
        if (variantCalls.length > 1) {
          extracted.variants = variantCalls;
        } else if (variantCalls.length === 1) {
          extracted.variants = variantCalls[0];
        }

        extractedComponents.push(extracted);
      }
    },
  });

  return extractedComponents;
}

/**
 * Extract a literal object from an AST ObjectExpression
 */
function extractObjectLiteral(node: t.ObjectExpression): Record<string, any> {
  const result: Record<string, any> = {};

  for (const prop of node.properties) {
    if (t.isObjectProperty(prop) && !prop.computed) {
      const key = t.isIdentifier(prop.key)
        ? prop.key.name
        : t.isStringLiteral(prop.key)
          ? prop.key.value
          : null;

      if (key) {
        const value = extractValue(prop.value);
        if (value !== undefined) {
          result[key] = value;
        }
      }
    }
  }

  return result;
}

/**
 * Extract a literal value from an AST node
 */
function extractValue(node: t.Node): any {
  if (t.isStringLiteral(node)) {
    return node.value;
  } else if (t.isNumericLiteral(node)) {
    return node.value;
  } else if (t.isBooleanLiteral(node)) {
    return node.value;
  } else if (t.isNullLiteral(node)) {
    return null;
  } else if (t.isObjectExpression(node)) {
    return extractObjectLiteral(node);
  } else if (t.isArrayExpression(node)) {
    return node.elements
      .filter((el): el is t.Expression | t.SpreadElement => el !== null)
      .map((el) => (t.isExpression(el) ? extractValue(el) : undefined))
      .filter((v) => v !== undefined);
  } else if (t.isTemplateLiteral(node) && node.expressions.length === 0) {
    // Simple template literal with no expressions
    return node.quasis[0].value.cooked;
  }

  // Return undefined for non-literal expressions
  return undefined;
}
