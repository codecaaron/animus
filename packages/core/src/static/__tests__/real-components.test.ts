import { extractStylesFromCode } from '../extractor';
import { CSSGenerator } from '../generator';
import {
  minimalColor,
  minimalSpace,
  minimalTypography,
  testTheme,
} from './testConfig';

describe('Real Component Patterns', () => {
  let generator: CSSGenerator;

  beforeEach(() => {
    generator = new CSSGenerator();
  });

  describe('Component Extraction', () => {
    it('extracts Logo component with complex styles', () => {
      const logoCode = `
        import { animus } from '@animus-ui/core';
        import { flow } from '../../animations/flow';

        export const Logo = animus
          .styles({
            width: 'max-content',
            fontSize: 30,
            m: 0,
            lineHeight: 'initial',
            fontFamily: 'logo',
            letterSpacing: '2px',
            gradient: 'flowX',
            backgroundSize: '300px 100px',
            animation: \` \${flow} 5s linear infinite\`,
            backgroundClip: 'text',
            WebkitTextFillColor: 'transparent',
            textShadow: 'logo',
            transition: 'text',
          })
          .states({
            link: {
              animation: 'none',
              '&:hover': {
                textShadow: 'logo-hover',
                animation: \` \${flow} 5s linear infinite\`,
              },
              '&:active': {
                textShadow: 'link-pressed',
              },
            },
          })
          .groups({ typography: true })
          .props({
            logoSize: {
              property: 'fontSize',
              scale: { xs: 28, sm: 32, md: 64, lg: 72, xl: 96, xxl: 128 },
            },
          })
          .asElement('h1');
      `;

      const extracted = extractStylesFromCode(logoCode);

      // Assert extraction structure instead of snapshot
      expect(extracted).toHaveLength(1);
      const logoStyle = extracted[0];

      // Verify component name
      expect(logoStyle.componentName).toBe('Logo');

      // Verify base styles
      expect(logoStyle.baseStyles).toEqual({
        width: 'max-content',
        fontSize: 30,
        m: 0,
        lineHeight: 'initial',
        fontFamily: 'logo',
        letterSpacing: '2px',
        gradient: 'flowX',
        backgroundSize: '300px 100px',
        backgroundClip: 'text',
        WebkitTextFillColor: 'transparent',
        textShadow: 'logo',
        transition: 'text',
      });

      // Verify states
      expect(logoStyle.states).toEqual({
        link: {
          animation: 'none',
          '&:hover': {
            textShadow: 'logo-hover',
          },
          '&:active': {
            textShadow: 'link-pressed',
          },
        },
      });

      // Verify groups
      expect(logoStyle.groups).toEqual(['typography']);

      // Verify props
      expect(logoStyle.props).toEqual({
        logoSize: {
          property: 'fontSize',
          scale: { xs: 28, sm: 32, md: 64, lg: 72, xl: 96, xxl: 128 },
        },
      });
    });

    it('extracts component with responsive values', () => {
      const code = `
        const HeaderSection = animus
          .styles({
            display: 'grid',
            gap: { _: 24, sm: 32 },
            flow: 'column',
            alignItems: 'center',
            flex: 1,
          })
          .variant({
            prop: 'direction',
            variants: {
              left: {},
              right: {
                justifyContent: 'end',
              },
            },
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);

      // Assert extraction structure instead of snapshot
      expect(extracted).toHaveLength(1);
      const headerStyle = extracted[0];

      // Verify component name
      expect(headerStyle.componentName).toBe('HeaderSection');

      // Verify base styles with responsive values
      expect(headerStyle.baseStyles).toEqual({
        display: 'grid',
        gap: { _: 24, sm: 32 },
        flow: 'column',
        alignItems: 'center',
        flex: 1,
      });

      // Verify variant structure
      expect(headerStyle.variants).toEqual({
        prop: 'direction',
        variants: {
          left: {},
          right: {
            justifyContent: 'end',
          },
        },
      });
    });
  });

  describe('CSS Generation', () => {
    it('generates CSS for component with all features', () => {
      const code = `
        const Button = animus
          .styles({
            padding: '8px 16px',
            backgroundColor: 'blue',
            color: 'white',
            border: 'none',
            cursor: 'pointer',
            '&:hover': {
              backgroundColor: 'darkblue',
            },
          })
          .variant({
            prop: 'variant',
            variants: {
              primary: {
                backgroundColor: 'darkblue',
              },
              secondary: {
                backgroundColor: 'gray',
                color: 'black',
              },
            },
          })
          .states({
            disabled: {
              opacity: 0.5,
              cursor: 'not-allowed',
            },
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('full-featured-component-css');
    });

    it('generates CSS with proper cascade ordering', () => {
      const code = `
        const Button = animus
          .styles({
            padding: '8px 16px',
            backgroundColor: 'blue',
          })
          .variant({
            prop: 'size',
            variants: {
              small: { padding: '4px 8px' },
              large: { padding: '12px 24px' },
            },
          })
          .variant({
            prop: 'variant',
            variants: {
              primary: { backgroundColor: 'darkblue' },
              secondary: { backgroundColor: 'gray' },
            },
          })
          .states({
            disabled: { opacity: 0.5 },
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(extracted[0]);
      expect(result.css).toMatchSnapshot('cascade-ordering-css');
    });
  });

  describe('Atomic Utilities Generation', () => {
    it('generates minimal atomic utilities for space group', () => {
      const code = `
        const Box = animus
          .styles({
            display: 'flex',
            backgroundColor: 'white',
          })
          .groups({
            space: true,
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(
        extracted[0],
        { space: minimalSpace },
        testTheme
      );
      expect(result.css).toMatchSnapshot('minimal-space-utilities');
    });

    it('generates atomic utilities for multiple groups', () => {
      const code = `
        const Card = animus
          .styles({
            borderRadius: 8,
            boxShadow: '0 2px 4px rgba(0,0,0,0.1)',
          })
          .groups({
            space: true,
            color: true,
            typography: true,
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const groupDefs = {
        space: minimalSpace,
        color: minimalColor,
        typography: minimalTypography,
      };
      const result = generator.generateFromExtracted(
        extracted[0],
        groupDefs,
        testTheme
      );
      expect(result.css).toMatchSnapshot('minimal-multiple-groups');
    });

    it('generates atomic utilities for custom props', () => {
      const code = `
        const Button = animus
          .styles({
            padding: '8px 16px',
            border: 'none',
            cursor: 'pointer',
          })
          .props({
            size: {
              property: 'fontSize',
              scale: 'fontSizes',
            },
            gap: {
              property: 'gap',
              scale: 'space',
            },
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(
        extracted[0],
        {},
        testTheme
      );
      expect(result.css).toMatchSnapshot('minimal-custom-props');
    });

    it('demonstrates shorthand expansion in atomic utilities', () => {
      const code = `
        const Container = animus
          .styles({
            display: 'grid',
          })
          .groups({
            space: true,
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      // Only include mx and py to show expansion clearly
      const groupDefs = {
        space: {
          mx: minimalSpace.mx,
          py: minimalSpace.py,
        },
      };
      const result = generator.generateFromExtracted(
        extracted[0],
        groupDefs,
        testTheme
      );
      expect(result.css).toMatchSnapshot('shorthand-expansion-demo');
    });

    it('places atomic utilities after component styles', () => {
      const code = `
        const Button = animus
          .styles({
            padding: '8px 16px',
            backgroundColor: 'blue',
            color: 'white',
          })
          .variant({
            prop: 'variant',
            variants: {
              primary: {
                backgroundColor: 'darkblue',
              },
            },
          })
          .states({
            disabled: {
              opacity: 0.5,
            },
          })
          .groups({
            space: true,
          })
          .asElement('button');
      `;

      const extracted = extractStylesFromCode(code);
      const result = generator.generateFromExtracted(
        extracted[0],
        { space: minimalSpace },
        testTheme
      );
      expect(result.css).toMatchSnapshot('cascade-with-utilities');
    });
  });

  describe('Theme Integration', () => {
    it('uses theme values for utilities generation', () => {
      const customTheme = {
        space: {
          xs: '0.25rem',
          sm: '0.5rem',
          md: '1rem',
        },
        colors: {
          brand: '#007bff',
          accent: '#6c757d',
        },
        fontSizes: {
          body: '16px',
          heading: '24px',
        },
      };

      const code = `
        const ThemedComponent = animus
          .styles({
            padding: '1rem',
          })
          .groups({
            space: true,
            color: true,
            typography: true,
          })
          .asElement('div');
      `;

      const extracted = extractStylesFromCode(code);
      const groupDefs = {
        space: minimalSpace,
        color: minimalColor,
        typography: minimalTypography,
      };
      const result = generator.generateFromExtracted(
        extracted[0],
        groupDefs,
        customTheme
      );
      expect(result.css).toMatchSnapshot('custom-theme-utilities');
    });
  });
});
