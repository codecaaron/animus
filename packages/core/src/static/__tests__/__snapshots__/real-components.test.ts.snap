// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Real Component Patterns Atomic Utilities Generation demonstrates shorthand expansion in atomic utilities: shorthand-expansion-demo 1`] = `
"/* Base Styles */
.animus-Container-c9r {
  display: grid;
}

/* Atomic Utilities from Groups */
/* Group: space */"
`;

exports[`Real Component Patterns Atomic Utilities Generation generates atomic utilities for custom props: minimal-custom-props 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 8px 16px;
  border: none;
  cursor: pointer;
}


/* Atomic Utilities from Custom Props */"
`;

exports[`Real Component Patterns Atomic Utilities Generation generates atomic utilities for multiple groups: minimal-multiple-groups 1`] = `
"/* Base Styles */
.animus-Card-c4d {
  border-radius: 8;
  box-shadow: 0 2px 4px rgba(0,0,0,0.1);
}

/* Atomic Utilities from Groups */
/* Group: space */
/* Group: color */
/* Group: typography */"
`;

exports[`Real Component Patterns Atomic Utilities Generation generates minimal atomic utilities for space group: minimal-space-utilities 1`] = `
"/* Base Styles */
.animus-Box-b3x {
  display: flex;
  background-color: var(--animus-colors-white);
}

/* Atomic Utilities from Groups */
/* Group: space */"
`;

exports[`Real Component Patterns Atomic Utilities Generation places atomic utilities after component styles: cascade-with-utilities 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 8px 16px;
  background-color: blue;
  color: var(--animus-colors-white);
}

/* Variant: variant="primary" */
.animus-Button-b6n-variant-primary {
  background-color: darkblue;
}

/* State: disabled */
.animus-Button-b6n-state-disabled {
  opacity: 0.5;
}

/* Atomic Utilities from Groups */
/* Group: space */"
`;

exports[`Real Component Patterns CSS Generation generates CSS for component with all features: full-featured-component-css 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 8px 16px;
  border: none;
  background-color: blue;
  color: white;
  cursor: pointer;
}
.animus-Button-b6n:hover {
  background-color: darkblue;
}

/* Variant: variant="primary" */
.animus-Button-b6n-variant-primary {
  background-color: darkblue;
}

/* Variant: variant="secondary" */
.animus-Button-b6n-variant-secondary {
  background-color: gray;
  color: black;
}

/* State: disabled */
.animus-Button-b6n-state-disabled {
  opacity: 0.5;
  cursor: not-allowed;
}"
`;

exports[`Real Component Patterns CSS Generation generates CSS with proper cascade ordering: cascade-ordering-css 1`] = `
"/* Base Styles */
.animus-Button-b6n {
  padding: 8px 16px;
  background-color: blue;
}

/* Variant: size="small" */
.animus-Button-b6n-size-small {
  padding: 4px 8px;
}

/* Variant: size="large" */
.animus-Button-b6n-size-large {
  padding: 12px 24px;
}

/* Variant: variant="primary" */
.animus-Button-b6n-variant-primary {
  background-color: darkblue;
}

/* Variant: variant="secondary" */
.animus-Button-b6n-variant-secondary {
  background-color: gray;
}

/* State: disabled */
.animus-Button-b6n-state-disabled {
  opacity: 0.5;
}"
`;

exports[`Real Component Patterns Theme Integration uses theme values for utilities generation: custom-theme-utilities 1`] = `
"/* Base Styles */
.animus-ThemedComponent-t15t {
  padding: 1rem;
}

/* Atomic Utilities from Groups */
/* Group: space */
/* Group: color */
/* Group: typography */"
`;
